File: 2965_find_missing_and_repeated_values.py
class Solution(object):
    def findMissingAndRepeatedValues(self, grid):
        n, N = len(grid), len(grid) * len(grid)
        S, S2 = N * (N + 1) // 2, N * (N + 1) * (2 * N + 1) // 6
        A = B = 0
        for r in grid:
            for v in r:
                A += v
                B += v * v
        d1, d2 = A - S, B - S2
        rep = (d1 + d2 // d1) // 2
        return [rep, rep - d1]

File: 2965_find_missing_and_repeated_values.java
class Solution {
    public int[] findMissingAndRepeatedValues(int[][] grid) {
        int n = grid.length, N = n * n;
        long sumExpected = (long) N * (N + 1) / 2;
        long sumSquareExpected = (long) N * (N + 1) * (2 * N + 1) / 6;
        long sumActual = 0, sumSquareActual = 0;

        for (int[] row : grid) {
            for (int num : row) {
                sumActual += num;
                sumSquareActual += (long) num * num;
            }
        }

        long diff1 = sumActual - sumExpected;  // a - b
        long diff2 = sumSquareActual - sumSquareExpected;  // a^2 - b^2

        int repeated = (int) ((diff1 + diff2 / diff1) / 2);
        int missing = (int) (repeated - diff1);

        return new int[]{repeated, missing};
    }
}
